src/Server/MySQLHandler.cpp

```diff
@@ -371,6 +384,79 @@ void MySQLHandler::comQuery(ReadBuffer & payload)
     }
 }
 
+void MySQLHandler::comStmtPrepare(DB::ReadBuffer & payload)
+{
+    if (prepared_statements_map.size() > 10000) /// Shouldn't happen in reality as COM_STMT_CLOSE cleans up the elements
+    {
+        LOG_ERROR(log, "Too many prepared statements");
+        current_prepared_statement_id = 0;
+        prepared_statements_map.clear();
+        packet_endpoint->sendPacket(ERRPacket(), true);
+        return;
+    }
+
+    String query;
+    readStringUntilEOF(query, payload);
+
+    uint32_t statement_id = current_prepared_statement_id;
+    if (current_prepared_statement_id == std::numeric_limits<uint32_t>::max())
+    {
+        current_prepared_statement_id = 0;
+    }
+    else
+    {
+        current_prepared_statement_id++;
+    }
+
+    // Key collisions should not happen here, as we remove the elements from the map with COM_STMT_CLOSE,
+    // and we have quite a big range of available identifiers with 32-bit unsigned integer
+    if (prepared_statements_map.contains(statement_id))
+    {
+        LOG_ERROR(
+            log,
+            "Failed to store a new statement `{}` with id {}; it is already taken by `{}`",
+            query,
+            statement_id,
+            prepared_statements_map.at(statement_id));
+        packet_endpoint->sendPacket(ERRPacket(), true);
+        return;
+    }
+
+    prepared_statements_map.emplace(statement_id, query);
+    packet_endpoint->sendPacket(PrepareStatementResponseOK(statement_id, 0, 0, 0), true);
+}
+
+void MySQLHandler::comStmtExecute(ReadBuffer & payload)
+{
+    uint32_t statement_id;
+    payload.readStrict(reinterpret_cast<char *>(&statement_id), 4);
+
+    if (!prepared_statements_map.contains(statement_id))
+    {
+        LOG_ERROR(log, "Could not find prepared statement with id {}", statement_id);
+        packet_endpoint->sendPacket(ERRPacket(), true);
+        return;
+    }
+
+    // Temporary workaround as we work only with queries that do not bind any parameters atm
+    ReadBufferFromString com_query_payload(prepared_statements_map.at(statement_id));
+    MySQLHandler::comQuery(com_query_payload, true);
+};
+
+void MySQLHandler::comStmtClose(ReadBuffer & payload)
+{
+    uint32_t statement_id;
+    payload.readStrict(reinterpret_cast<char *>(&statement_id), 4);
+
+    if (prepared_statements_map.contains(statement_id))
```

Comment:
Does access to the map need synchronization? (i.e. some mutex?). I mean even if it is local to a single MySQLHandler, I would not risk races (due to network package delays or whatever).

Reply:
This is a good one. I assumed that we should not have race conditions here due to 1 MySQLHandler/thread. Will wrap it 👍